Hexastore/Query/ObjectQueryExecutor.cs

```diff
@@ -190,66 +198,140 @@ private IEnumerable<Triple> CreateConstraint(IStoreGraph graph, string key, Quer
 
         private IEnumerable<Triple> ApplyOutgoing(IEnumerable<Triple> source, IStoreGraph graph, LinkQuery link)
         {
-            if (link == null) {
+            if (link == null)
+            {
                 return source;
             }
 
-            if (string.IsNullOrEmpty(link.Path)) {
+            if (string.IsNullOrEmpty(link.Path))
+            {
                 throw _storeErrors.PathEmpty;
             }
 
             var paths = link.Path.Split(LinkDelimiterArray);
-
+            var matchingTargets = new HashSet<string>();
+            var previouslySeenTargets = new HashSet<string>();
             var matched = source.Where(x =>
             {
-                IEnumerable<string> targets;
+                IDictionary<string, ISet<string>> targets;
```

Comment:
nit: use `Dictionary<string, HashSet<string>>  Using the actual types is better for perf, and since this isn't passed externally there is no benefit to using the interfaces.

Reply:
Got sidetracked. Thanks for the feedback, updated in https://github.com/angshuman/hexadb/pull/18/commits/b0701ade67eed69ff53ac089797816838cdbbd68